src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonClass.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Interpreter.Ast {
+    class AstPythonClass : AstPythonType, IPythonClass {
+        protected static readonly IPythonModule NoDeclModule = new AstPythonModule();
+
+        private readonly object _lock = new object();
+
+        private IReadOnlyList<IPythonType> _mro;
+        private AsyncLocal<bool> _isProcessing = new AsyncLocal<bool>();
+
+        public AstPythonClass(
+            ClassDefinition classDefinition,
+            IPythonModule declaringModule,
+            string doc,
+            ILocationInfo loc,
+            BuiltinTypeId builtinTypeId = BuiltinTypeId.Type
+        ) : base(classDefinition.Name, declaringModule, doc, loc, builtinTypeId, false) {
+            ClassDefinition = classDefinition;
+        }
+
+        internal AstPythonClass(string name) : base(name, BuiltinTypeId.Type) { }
+
+        #region IPythonType
+        public override PythonMemberType MemberType => PythonMemberType.Class;
+
+        public override IMember GetMember(IModuleContext context, string name) {
+            IMember member;
+            lock (_lock) {
```

Comment:
Is this a new code or was it moved from somewhere?

Reply:
Moved. It was all in python type which I split into type and class so SetBases don't apply to basic types.